################################################################################
SHORT DESCRIPTION: 
################################################################################
Treat a vanishing backend as a request to close the front-end.

################################################################################
LONG DESCRIPTION: 
################################################################################
Usual scenario would be dom0 releaving netback responsibility to NDVM ending up
to rmmod its backend module. Xenstore node would disapear and Xenbus switch to
unknown state. We might as well consider it, on the frontend side (e.g, in
UIVM) to be a close request, so the guest will be able to reconnect with the
new backend in the service VM.

################################################################################
CHANGELOG 
################################################################################
Original author: Steve Meisner <steve.meisner@citrix.com>
Port to 3.18: Eric Chanudet <chanudete@ainfosec.com>

################################################################################
REMOVAL 
################################################################################
Supporting backend relocation would require some work in the current protocol
without this patch.

################################################################################
UPSTREAM PLAN
################################################################################
None, this is an OpenXT work-around.

################################################################################
INTERNAL DEPENDENCIES 
################################################################################
Service VM PV backend.

################################################################################
PATCHES 
################################################################################
Index: linux-4.1.13/drivers/net/xen-netfront.c
===================================================================
--- linux-4.1.13.orig/drivers/net/xen-netfront.c	2015-11-17 15:38:56.673137336 +0100
+++ linux-4.1.13/drivers/net/xen-netfront.c	2015-11-17 15:39:05.393034366 +0100
@@ -1975,15 +1975,26 @@
 	dev_dbg(&dev->dev, "%s\n", xenbus_strstate(backend_state));
 
 	switch (backend_state) {
+	case XenbusStateUnknown:
+		/* if the backend vanishes from xenstore, close frontend */
+		if (!xenbus_exists(XBT_NIL, dev->otherend, "") &&
+		    (dev->state != XenbusStateClosed)) {
+			dev_warn(&dev->dev,
+				"backend vanished, closing frontend\n");
+			xenbus_frontend_closed(dev);
+			xennet_disconnect_backend(np);
+		}
+		break;
 	case XenbusStateInitialising:
 	case XenbusStateInitialised:
 	case XenbusStateReconfiguring:
 	case XenbusStateReconfigured:
-	case XenbusStateUnknown:
 		break;
 
 	case XenbusStateInitWait:
-		if (dev->state != XenbusStateInitialising)
+		/* allow reconnect if our state is either initialising, or closed */
+		if (dev->state != XenbusStateInitialising &&
+		    dev->state != XenbusStateClosed)
 			break;
 		if (xennet_connect(netdev) != 0)
 			break;
